### PR TITLE
Oop factory

### PR DESCRIPTION
**Pull request by nnhw**
I've created basic factory interface and realization. I'll continue improvements, just need to verify an overall direction

### CODE REVIEW

File: lib/includes/ILinkFactory.h [hunk 1]
```diff
@@ -0,0 +1,16 @@
+#ifndef LOODSMAN_ILINKFACTORY_H
+#define LOODSMAN_ILINKFACTORY_H
```
Comment (MishkaRogachev) [inline lib/includes/ILinkFactory.h:2]: No define prefix

File: lib/includes/ILinkFactory.h [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+#include "i_link.h"
+
+namespace loodsman
+{
+class ILinkFactory
+{
+public:
+    virtual ~ILinkFactory() = default;
+
+    virtual int create(ILink* link) = 0;
```
Comment (MishkaRogachev) [inline lib/includes/ILinkFactory.h:13]: virtual ILink* create() = 0;

File: lib/includes/UdpLinkFactory.h [hunk 3]
```diff
@@ -0,0 +1,29 @@
+#include "ILinkFactory.h"
```
Comment (MishkaRogachev) [inline lib/includes/UdpLinkFactory.h:1]: files in snake_case

File: CMakeLists.txt [hunk 4]
```diff
@@ -3,8 +3,8 @@ cmake_minimum_required(VERSION 3.7.2)
 cmake_policy(VERSION 3.4.1)
 
 if (NOT DEFINED CMAKE_CXX_STANDARD)
-  set(CMAKE_CXX_STANDARD 17)
-endif()
+    set(CMAKE_CXX_STANDARD 17)
+endif ()
```
Comment (MishkaRogachev) [inline CMakeLists.txt:7]: No spaces before braces in CMake

File: Testing/Temporary/CTestCostData.txt [hunk 5]
```diff
@@ -0,0 +1 @@
+---
```
Comment (MishkaRogachev) [inline Testing/Temporary/CTestCostData.txt:1]: Test artifacts?

File: lib/includes/UdpLinkFactory.h [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+#include "i_link.h"
+#include <string_view>
```
Comment (MishkaRogachev) [inline lib/includes/UdpLinkFactory.h:3]: Includes order:

#include "i_link_factory.h"

#include <string_view>

#include "i_link.h"

File: lib/includes/UdpLinkFactory.h [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+
+#ifndef LOODSMAN_UDPLINKFACTORY_H
+#define LOODSMAN_UDPLINKFACTORY_H
+
+namespace loodsman
+{
+class UdpLinkFactory : public ILinkFactory
+{
+public:
+    UdpLinkFactory(int local_port);
```
Comment (MishkaRogachev) [inline lib/includes/UdpLinkFactory.h:13]: camelCase varibles

File: lib/includes/UdpLinkFactory.h [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+
```
Comment (MishkaRogachev) [inline lib/includes/UdpLinkFactory.h:14]: no spaces between constructors

File: lib/includes/UdpLinkFactory.h [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+    UdpLinkFactory(int local_port, std::string local_address);
```
Comment (MishkaRogachev) [inline lib/includes/UdpLinkFactory.h:15]: string_view in arguments

File: lib/includes/UdpLinkFactory.h [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+
+    UdpLinkFactory(int local_port, std::string local_address, int remote_port,
```
Comment (MishkaRogachev) [inline lib/includes/UdpLinkFactory.h:17]: Main (full) constructor must be first

File: lib/sources/UdpLinkFactory.cpp [hunk 11]
```diff
@@ -0,0 +1,53 @@
+#include "UdpLinkFactory.h"
```
Comment (MishkaRogachev) [inline lib/sources/UdpLinkFactory.cpp:1]: includes order

File: lib/sources/UdpLinkFactory.cpp [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+#include "ILinkFactory.h"
+#include "i_link.h"
+#include "udp_link.h"
+#include "utils.h"
+#include <string_view>
+
+using namespace loodsman;
+
+UdpLinkFactory::UdpLinkFactory(int local_port) :
```
Comment (MishkaRogachev) [inline lib/sources/UdpLinkFactory.cpp:10]: Use main constructor in others

File: lib/sources/UdpLinkFactory.cpp [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+        debug_print("boost system error");
+        return -1;
+    }
+    catch (...)
+    {
+        debug_print("Generic error");
+        return -1;
+    }
+
+    return 0;
```
Comment (MishkaRogachev) [inline lib/sources/UdpLinkFactory.cpp:52]: return link object

File: lib/sources/link_factory.cpp [hunk 14]
```diff
@@ -6,33 +6,34 @@
 
 using namespace loodsman;
 
-int loodsman::factory(LinkPtr& linkOut ,link_type type, int local_port, std::string local_address, int remote_port,
-                           std::string remote_address)
+int loodsman::factory(LinkPtr& linkOut, link_type type, int local_port, std::string local_address,
```
Comment (MishkaRogachev) [inline lib/sources/link_factory.cpp:9]: return link object

Comment (MishkaRogachev) [reply lib/sources/link_factory.cpp:9]: Turn it to class. LinksFactory or LinksCreator

Comment (nnhw) [reply lib/sources/link_factory.cpp:9]: I'll delete old factory after finishing this one

**Pull request merged**